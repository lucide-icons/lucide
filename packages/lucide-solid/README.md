# Lucide Solid

Implementation of the lucide icon library for solid applications.

> What is lucide? Read it [here](https://github.com/lucide-icons/lucide#what-is-lucide).

## Installation

```sh
yarn add lucide-solid
```

or

```sh
npm install lucide-solid
```

## How to use

It's build with ESmodules so it's completely tree-shakable.
Each icon can be imported as a solid component.

### Example

You can pass additional props to adjust the icon.

```js
import { Camera } from 'lucide-solid';
// Returns SolidComponent

// Usage
const App = () => {
  return <Camera color="red" size={48} />;
};

export default App;
```

### Props

| name          | type     | default      |
| ------------- | -------- | ------------ |
| `size`        | _Number_ | 24           |
| `color`       | _String_ | currentColor |
| `strokeWidth` | _Number_ | 2            |

### Custom props / svg attributes

You can also pass custom props that will be added in the as attributes. With that you can modify the icons look by passing svg attributes.

```js
// Usage
const App = () => {
  return <Camera fill="red" stroke-linejoin="bevel" />;
};
```

### One generic icon component

It is possible to create one generic icon component to load icons.

> :warning: Example below importing all EsModules, caution using this example, not recommended when you using bundlers, your application build size will grow strongly.

#### Icon Component Example

```tsx
import * as icons from 'lucide-solid';
import type { LucideProps } from 'lucide-solid';
import { splitProps } from 'solid-js';
import { Dynamic } from 'solid-js/web';

const Icon = (props: { name: keyof typeof icons } & LucideProps) => {
  const [local, others] = splitProps(props, ["name"]);

  return <Dynamic component={icons[local.name]} {...others} />
};

export default Icon;
```
