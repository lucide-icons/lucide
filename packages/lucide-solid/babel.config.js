const mainConfig = require('../../babel.config');

module.exports = {
  presets: ['solid'],
  env: mainConfig.env,
};
