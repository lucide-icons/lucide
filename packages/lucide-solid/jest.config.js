module.exports = {
  verbose: true,
  // preset: 'solid-jest/preset/browser',
  roots: ['<rootDir>/src/', '<rootDir>/tests/'],
  moduleFileExtensions: ['js'],
  transformIgnorePatterns: [`/node_modules`],
  transform: {
    '^.+\\.js$': 'babel-jest',
  },
};
