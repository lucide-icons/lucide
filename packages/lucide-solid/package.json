{
  "name": "lucide-solid",
  "description": "A Lucide icon library package for Solid applications",
  "version": "0.88.0-beta.3",
  "license": "ISC",
  "homepage": "https://lucide.dev",
  "bugs": "https://github.com/lucide-icons/lucide/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/lucide-icons/lucide.git",
    "directory": "packages/lucide-solid"
  },
  "author": "Eric Fennis",
  "amdName": "lucide-solid",
  "main": "dist/cjs/lucide-solid.js",
  "main:umd": "dist/umd/lucide-solid.js",
  "module": "dist/esm/lucide-solid.js",
  "unpkg": "dist/umd/lucide-solid.min.js",
  "typings": "dist/lucide-solid.d.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "pnpm clean && pnpm copy:license && pnpm build:icons && pnpm build:bundles && pnpm build:types",
    "copy:license": "cp ../../LICENSE ./LICENSE",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.js",
    "build:icons": "node ../../scripts/buildIcons.mjs --output=./src --templateSrc=./scripts/exportTemplate.mjs --renderUniqueKey --iconFileExtension=.ts --exportFileName=index.ts",
    "build:types": "node ./scripts/buildTypes.mjs",
    "build:bundles": "rollup -c ./rollup.config.mjs",
    "test": "vitest run",
    "version": "pnpm version --git-tag-version=false"
  },
  "devDependencies": {
    "@lucide/rollup-plugins": "workspace:*",
    "@testing-library/jest-dom": "^5.16.5",
    "babel-preset-solid": "^1.5.4",
    "jsdom": "^20.0.0",
    "rollup": "^3.5.1",
    "solid-js": "^1.4.7",
    "solid-testing-library": "^0.3.0",
    "vite": "^3.2.4",
    "vite-plugin-solid": "^2.3.0",
    "vitest": "^0.23.2"
  },
  "peerDependencies": {
    "solid-js": "^1.4.7"
  }
}
