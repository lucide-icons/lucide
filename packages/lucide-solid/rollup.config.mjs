import plugins from '@lucide/rollup-plugins';
import pkg from './package.json' assert { type: 'json' };

const packageName = 'LucideSolid';
const outputFileName = 'lucide-solid';
const outputDir = 'dist';
const inputs = [`src/lucide-solid.ts`];
const bundles = [
  {
    format: 'umd',
    inputs,
    outputDir,
    minify: true,
  },
  {
    format: 'umd',
    inputs,
    outputDir,
  },
  {
    format: 'cjs',
    inputs,
    outputDir,
  },
  {
    format: 'es',
    inputs,
    outputDir,
  },
  {
    format: 'esm',
    inputs,
    outputDir,
    preserveModules: true,
  },
];

const configs = bundles
  .map(({ inputs, outputDir, format, minify, preserveModules }) =>
    inputs.map(input => ({
      input,
      plugins: plugins(pkg, minify),
      external: ['solid-js', 'solid-js/h'],
      output: {
        name: packageName,
        ...(preserveModules
          ? {
              dir: `${outputDir}/${format}`,
            }
          : {
              file: `${outputDir}/${format}/${outputFileName}${minify ? '.min' : ''}.js`,
            }),
        preserveModules,
        format,
        sourcemap: true,
        globals: {
          'solid-js': 'solid-js',
          'solid-js/h': 'solid-js/h',
        },
      },
    })),
  )
  .flat();

export default configs;
