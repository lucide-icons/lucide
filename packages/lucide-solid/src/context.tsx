import { createContext, splitProps, type JSXElement } from 'solid-js';

export const LucideContext = createContext<{
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}>({
  size: 24,
  fill: 'none',
  color: 'currentColor',
  strokeWidth: 2,
  absoluteStrokeWidth: false,
});

interface LucideProviderProps {
  children: JSXElement;
  size?: number;
  fill?: string;
  color?: string;
  strokeWidth?: number;
  absoluteStrokeWidth?: boolean;
}

export function LucideProvider(props: LucideProviderProps) {
  const [value, rest] = splitProps(props, ['size', 'color', 'strokeWidth', 'absoluteStrokeWidth']);

  return <LucideContext.Provider value={value}>{rest.children}</LucideContext.Provider>;
}
