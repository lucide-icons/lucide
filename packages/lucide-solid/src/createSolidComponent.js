import h from 'solid-js/h';
import { mergeProps } from 'solid-js';
import defaultAttributes from './defaultAttributes';

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = string => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

export default (iconName, iconNode) => {
  const Component = ({
    color = 'currentColor',
    size = 24,
    strokeWidth = 2,
    children = [],
    ...rest
  }) =>
    h(
      'svg',
      mergeProps(
        defaultAttributes,
        {
          width: size,
          height: size,
          stroke: color,
          'stroke-width': strokeWidth,
          classList: {
            lucide: true,
            [`lucide-${toKebabCase(iconName)}`]: true,
          },
        },
        rest,
      ),
      [...iconNode.map(([tag, attrs]) => h(tag, attrs)), children],
    );

  Component.displayName = `${iconName}`;

  return Component;
};
