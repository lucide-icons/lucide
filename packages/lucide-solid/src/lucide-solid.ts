/**
 * lucide-solid {{version}} - ISC
 */

export * from './icons';
export * as icons from './icons';
export * from './aliases';
