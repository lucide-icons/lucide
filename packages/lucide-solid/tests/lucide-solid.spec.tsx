import { describe, it, expect } from 'vitest';
import { cleanup, render } from 'solid-testing-library'
import { Edit2, Grid, Pen } from '../src/lucide-solid'

describe('Using lucide icon components', () => {
  it('should render an component', () => {
    const { container } = render(() => <Grid /> );

    expect( container.innerHTML ).toMatchSnapshot();
    cleanup()
  });

  it('should adjust the size, stroke color and stroke width', async () => {
    const testId = 'grid-icon';
    const { container, getByTestId } = render(() =>
      <Grid
        data-testid={testId}
        size={48}
        color="red"
        strokeWidth={4}
      />,
    );

    const { attributes } = await getByTestId(testId) as unknown as{ attributes: Record<string, { value: string }>};
    expect(attributes.stroke.value).toBe('red');
    expect(attributes.width.value).toBe('48');
    expect(attributes.height.value).toBe('48');
    expect(attributes['stroke-width'].value).toBe('4');
    expect( container.innerHTML ).toMatchSnapshot();
    cleanup()
  });

  it('should render the alias icon', () => {
    const testId = 'pen-icon';
    const { container } = render(() =>
      <Pen
        data-testid={testId}
        size={48}
        stroke="red"
        strokeWidth={4}
      />,
    );

    const PenIconRenderedHTML = container.innerHTML

    cleanup()

    const { container: Edit2Container } = render(() =>
      <Edit2
        data-testid={testId}
        size={48}
        stroke="red"
        strokeWidth={4}
      />,
    );

    expect(PenIconRenderedHTML).toBe(Edit2Container.innerHTML)
    cleanup()
  });


  it('should not scale the strokeWidth when absoluteStrokeWidth is set', () => {
    const testId = 'grid-icon';
    const { container, getByTestId } = render(() =>
      <Grid
        data-testid={testId}
        size={48}
        color="red"
        absoluteStrokeWidth
      />,
    );

    const { attributes } = getByTestId(testId) as unknown as{ attributes: Record<string, { value: string }>};
    expect(attributes.stroke.value).toBe('red');
    expect(attributes.width.value).toBe('48');
    expect(attributes.height.value).toBe('48');
    expect(attributes['stroke-width'].value).toBe('1');

    expect( container.innerHTML ).toMatchSnapshot();
    cleanup()
  });
})
