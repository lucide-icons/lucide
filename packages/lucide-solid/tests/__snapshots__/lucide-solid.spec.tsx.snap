// Vitest Snapshot v1

exports[`Using lucide icon components > should adjust the size, stroke color and stroke width 1`] = `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"red\\" stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"lucide lucide-grid \\" data-testid=\\"grid-icon\\"><rect x=\\"3\\" y=\\"3\\" width=\\"18\\" height=\\"18\\" rx=\\"2\\" ry=\\"2\\" key=\\"maln0c\\"></rect><line x1=\\"3\\" y1=\\"9\\" x2=\\"21\\" y2=\\"9\\" key=\\"1uch6j\\"></line><line x1=\\"3\\" y1=\\"15\\" x2=\\"21\\" y2=\\"15\\" key=\\"1xojw2\\"></line><line x1=\\"9\\" y1=\\"3\\" x2=\\"9\\" y2=\\"21\\" key=\\"nvcl17\\"></line><line x1=\\"15\\" y1=\\"3\\" x2=\\"15\\" y2=\\"21\\" key=\\"jcv93v\\"></line></svg>"`;

exports[`Using lucide icon components > should render an component 1`] = `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"lucide lucide-grid \\"><rect x=\\"3\\" y=\\"3\\" width=\\"18\\" height=\\"18\\" rx=\\"2\\" ry=\\"2\\" key=\\"maln0c\\"></rect><line x1=\\"3\\" y1=\\"9\\" x2=\\"21\\" y2=\\"9\\" key=\\"1uch6j\\"></line><line x1=\\"3\\" y1=\\"15\\" x2=\\"21\\" y2=\\"15\\" key=\\"1xojw2\\"></line><line x1=\\"9\\" y1=\\"3\\" x2=\\"9\\" y2=\\"21\\" key=\\"nvcl17\\"></line><line x1=\\"15\\" y1=\\"3\\" x2=\\"15\\" y2=\\"21\\" key=\\"jcv93v\\"></line></svg>"`;
